src/rust/multi.mangadex/src/lib.rs

```diff
@@ -1,311 +1,282 @@
 #![no_std]
+#![feature(let_chains)]
+extern crate alloc;
+mod helper;
+mod parser;
 use aidoku::{
-	prelude::*, error::Result, std::String, std::Vec, std::ObjectRef, std::net::Request, std::net::HttpMethod,
-	Filter, FilterType, Listing, Manga, MangaPageResult, Chapter, Page, DeepLink,
-	std::defaults::defaults_get,
+	error::*,
+	prelude::*,
+	std::{
+		defaults::{defaults_get, defaults_set},
+		net::{HttpMethod, Request},
+		ArrayRef, ObjectRef, String, StringRef, Vec,
+	},
+	Chapter, DeepLink, Filter, FilterType, Listing, Manga, MangaPageResult, Page,
 };
+use alloc::borrow::ToOwned;
+use helper::*;
 
-mod parser;
-
-fn urlencode(string: String) -> String {
-	let mut result: Vec<u8> = Vec::with_capacity(string.len() * 3);
-	let hex = "0123456789abcdef".as_bytes();
-	let bytes = string.as_bytes();
-	
-	for byte in bytes {
-		let curr = *byte;
-		if (b'a' <= curr && curr <= b'z')
-			|| (b'A' <= curr && curr <= b'Z')
-			|| (b'0' <= curr && curr <= b'9') {
-				result.push(curr);
-		} else {
-			result.push(b'%');
-			result.push(hex[curr as usize >> 4]);
-			result.push(hex[curr as usize & 15]);
-		}
-	}
-
-	String::from_utf8(result).unwrap_or(String::new())
+#[link(wasm_import_module = "net")]
+extern "C" {
+	fn set_rate_limit(rate_limit: i32);
+	fn set_rate_limit_period(period: i32);
 }
 
-fn i32_to_string(mut integer: i32) -> String {
-	if integer == 0 {
-		return String::from("0");
-	}
-	let mut string = String::with_capacity(11);
-	let pos = if integer < 0 {
-		string.insert(0, '-');
-		1
-	} else {
-		0
-	};
-	while integer != 0 {
-		let mut digit = integer % 10;
-		if pos == 1 {
-			digit *= -1;
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn initialize() {
+	set_rate_limit(3);
+	set_rate_limit_period(1);
+
+	for key in ["blockedGroups", "blockedUploaders"] {
+		let arrkey = key.to_owned() + "Array";
+		if defaults_get(&arrkey).as_array().is_err() {
+			handle_notification(String::from(key));
 		}
-		string.insert(pos, char::from_u32((digit as u32) + ('0' as u32)).unwrap());
-		integer /= 10;
 	}
-	return string;
 }
 
 #[get_manga_list]
 fn get_manga_list(filters: Vec<Filter>, page: i32) -> Result<MangaPageResult> {
 	let offset = (page - 1) * 20;
-	let mut url = String::from("https://api.mangadex.org/manga/?includes[]=cover_art&limit=20&offset=");
-	url.push_str(&i32_to_string(offset));
+	let mut url = String::from(
+		"https://api.mangadex.org/manga/?includes[]=cover_art\
+		&limit=20\
+		&offset=",
+	) + itoa::Buffer::new().format(offset);
```

Comment:
any reason to use itoa? I don't think i32_to_string was bad

Reply:
It's [fast](https://docs.rs/itoa/latest/itoa/#performance-lower-is-better) and it doesn't take up that much space so I figured why not eh